File: steit-csharp/src/State/Path.cs
using System;
using System.Text;

namespace Steit.State {
    public sealed class Path {
        public static Path Root = new Path(null, null);

        // public Path? Parent { get; private set; }
        public Path Parent { get; private set; }
        public UInt32? Tag { get; private set; }

        // private Path(Path? parent, UInt32? tag) {
        private Path(Path parent, UInt32? tag) {
            this.Parent = parent;
            this.Tag = tag;
        }

        public Path GetNested(UInt32 tag) {
            return new Path(this, tag);
        }

        public override string ToString() {
            var builder = new StringBuilder();
            this.ToString(builder);
            return builder.ToString();
        }

        private void ToString(StringBuilder builder) {
            if (this.Parent != null) {
                this.Parent.ToString(builder);
                builder.AppendFormat("/{0}", this.Tag);
            }
        }
    }
}

File: steit-go/pkg/path/path.go
package path

import (
	"fmt"
	"strings"
)

type Path struct {
	parent *Path
	tag    *uint32
}

var (
	Root = &Path{nil, nil}
)

func (p *Path) GetParent() *Path {
	return p.parent
}

func (p *Path) GetTag() *uint32 {
	return p.tag
}

func (p *Path) GetNested(tag uint32) *Path {
	return &Path{parent: p, tag: &tag}
}

func (p *Path) String() string {
	var builder strings.Builder
	p.buildPath(&builder)
	return builder.String()
}

func (p *Path) buildPath(builder *strings.Builder) {
	if p.parent != nil {
		p.parent.buildPath(builder)
		builder.WriteString(fmt.Sprintf("/%d", *p.tag))
	}
}
